experimental/packages/opentelemetry-sdk-metrics-base/src/MeterProvider.ts

```
@@ -0,0 +1,170 @@
+            api.diag.warn('A shutdown MeterProvider cannot provide a Meter')
+            return metrics.NOOP_METER;
+        }
+
+        // Spec leaves it unspecified if creating a meter with duplicate
+        // name/version returns the same meter. We create a new one here
+        // for simplicity. This may change in the future.
+        return new Meter(this, { name, version }, options.schemaUrl);
+    }
+
+    addMetricReader(metricReader: MetricReader) {
+        this._metricReaders.push(metricReader);
+    }
+
+    addView(view: View) {
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view
+        this._views.push(view);
+    }
+
+    async shutdown(): Promise<void> {
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#shutdown
+
+        if (this._shutdown) {
+            api.diag.warn('shutdown may only be called once per MeterProvider');
+            return;
+        }
+
+        // TODO add a timeout - spec leaves it up the the SDK if this is configurable
+        this._shutdown = true;
+
+        // Shut down all exporters and readers.
+        // Throw the first error and log all others.
+        let err: unknown;
+        for (const exporter of this._metricExporters) {
+            try {
+                await exporter.shutdown();
+            } catch (e) {
+                if (e instanceof Error) {
+                    api.diag.error(`Error shutting down: ${e.message}`)
+                }
+                err = err || e;
+            }
+        }
+
+        for (const reader of this._metricReaders) {
+            try {
+                await reader.shutdown();
+            } catch (e) {
+                if (e instanceof Error) {
+                    api.diag.error(`Error shutting down: ${e.message}`)
+                }
+                err = err || e;
+            }
+        }
+
+        if (err != null) {
```

shouldn't we never throw ? not related but we can technically throw anything (including null & undefined) so i would prefer to check `err instanceof Error` here

---

Spec says:

>Shutdown SHOULD provide a way to let the caller know whether it succeeded, failed or timed out.

I suppose we could do that with return value instead of throw?

---

Well if we can throw in the sdk i suppose it make more sense to do it, if not i guess we should return an error yeah

---

I'll leave them as diag logs for now. That's what we do most other places and this isn't a startup/config issue so we can't "fail early"